bool b1
bool b2
bool b3
bool b4
bool b5
bool b6
bool b7
bool b8